### PR TITLE
Ab#80203 remove graphql from export functions

### PR DESCRIPTION
**Pull request by MwanPygmay**
# Description

Remove graphql from the export to improve performance. Resources, resource, choices by url, calculated fields, and reference data  (even nested) are supported.

## Useful links

- [Please insert link to ticket](https://dev.azure.com/WHOHQ/EMSSAFE/_workitems/edit/80203/)
- [Please insert link to front-end branch if any](https://github.com/ReliefApplications/ems-frontend/pull/2117)

## Type of change

- [x] Improvement (refactor or addition to existing functionality)

# How Has This Been Tested?

Created resource with 11000 records, with a calculated field. Tested performance for this resource.
Tested csv, xlsx, resource and resources fields.

## Screenshots

Performance with and without graphql for 10 takes:
Screenshot 2023-12-01 at 13 18 33

# Checklist:

( * == Mandatory ) 

- [x] * I have set myself as assignee of the pull request
- [x] * My code follows the style guidelines of this project
- [x] * Linting does not generate new warnings
- [x] * I have performed a self-review of my own code
- [x] * I have put the ticket for review, adding the oort-backend team to the list of reviewers
- [x] * I have commented my code, particularly in hard-to-understand areas
- [x] * I have put JSDoc comment in all required places
- [x] * My changes generate no new warnings
- [x] * I have included screenshots describing my changes if relevant
- [x] * I have selected labels in the Pull Request, according to the changes with code brings
- [ ] I have made corresponding changes to the documentation ( if required )
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/utils/files/exportBatch.ts [hunk 1]
```diff
@@ -291,146 +254,165 @@ const writeRowsXlsx = (
...
+          params.timeZone
+        ) as any)
+      )
+    );
 
-    offset += batchSize;
-  } while (offset < totalCount);
-};
+  const records = await Record.aggregate<Record>(pipeline as any);
+  console.log(records.slice(0, 5));
```
Comment (unai-reliefapp) [inline src/utils/files/exportBatch.ts:311]: image
```suggestion
```

File: src/utils/files/exportBatch.ts [hunk 2]
```diff
@@ -291,146 +254,165 @@ const writeRowsXlsx = (
...
+      (await getChoices(column.meta.field, authorization)).map((country) => [
+        country.value,
+        country.text,
+      ])
+    );
+    records.forEach((record) => {
+      record[column.field] = choices[record[column.field]];
+    });
+  }
+  console.log(records.slice(0, 5));
```
Comment (unai-reliefapp) [inline src/utils/files/exportBatch.ts:327]: image

```suggestion
```

**Review by unai-reliefapp — CHANGES_REQUESTED** [review]

**Review by estelafs — APPROVED** [review]
Everything is working as expected

**Comment by AntoineRelief** [comment]
made some improvements to it:
- better performance, indeed, a for loop on items fetched for related resources would take too much time ( doing one query per record is more efficient it seems )
- use datastream
- use existing function to write values from columns meta

to be done:
- check if working when using reference data questions, I had to comment this part to make some improvements
- not working for some resource(s) links. If the resource(s) question is part of the resource you export, it is working fine, but it you are using a query where the link is coming from the other resource ( question using currently exported resource ), it won't work. We should check how we did that when using graphql ( in the src / utils / schema / resolvers / entity file ) as it is working fine there

**Comment by MwanPygmay** [comment]
- Fix some bugs, notably for resource/resources/reference data questions
- Changed a bit @LenaelleRelief's logic to get the parent resource(s) directly from the front (see front PR) => this also fetches the children resources (not resource), but did not find how to avoid it, and it causes no bug, just some optimization. This avoids some problems, Lenaelle searched for parent resource(s) by field and it could lead to some problem if several resources had the same field. The new method should be sure.

**Comment by AntoineRelief** [comment]
now using this pr:
https://github.com/ReliefApplications/ems-backend/pull/961

**Pull request closed without merging**